zstash/globus.py

```diff
@@ -49,18 +78,22 @@ def check_endpoint_version_5(ep_id):
     return False
 
 
-def submit_transfer_with_checks(transfer_data):
+def submit_transfer_with_checks(globus_info: GlobusInfo):
+    if not globus_info.transfer_client:
+        raise ValueError("transfer_client is undefined")
     try:
-        task = transfer_client.submit_transfer(transfer_data)
+        task = globus_info.transfer_client.submit_transfer(globus_info.transfer_data)
     except TransferAPIError as err:
         if err.info.consent_required:
             scopes = "urn:globus:auth:scope:transfer.api.globus.org:all["
-            for ep_id in [remote_endpoint, local_endpoint]:
-                if check_endpoint_version_5(ep_id):
+            for ep_id in [globus_info.remote_endpoint, globus_info.local_endpoint]:
+                if check_endpoint_version_5(globus_info, ep_id):
                     scopes += f" *https://auth.globus.org/scopes/{ep_id}/data_access"
             scopes += " ]"
```

Comment:
We can't do this scope expansion until after `transfer_client` gets set. It appears the two authentications are 1) initiating the native client and 2) adding the scopes of the local and remote endpoints to it.

Reply:
@forsyth2 Although I'd never looked very hard, I had assume the first involved Globus recognizing the account as a valid Globus user, and then conducting ep-authentication/scoping.  Surprised there are not 3 of these (globus, ep-1, ep-2)

I have gotten different results (even in the globus web UI) by when authenticating to endpoints in a different order.